File: java/tengi-core/src/main/java/com/noctarius/tengi/serialization/codec/Decoder.java
package com.noctarius.tengi.serialization.codec;

import com.noctarius.tengi.buffer.ReadableMemoryBuffer;

public interface Decoder {

    int readBytes(byte[] bytes);

    int readBytes(byte[] bytes, int offset, int length);

    boolean readBoolean();

    boolean[] readBitSet();

    byte readByte();

    short readUnsignedByte();

    short readShort();

    char readChar();

    int readInt32();

    int readCompressedInt32();

    long readInt64();

    long readCompressedInt64();

    float readFloat();

    double readDouble();

    String readString();

    <O> O readObject()
            throws Exception;

    <O> O readNullableObject()
            throws Exception;

    ReadableMemoryBuffer getReadableMemoryBuffer();

}

File: javascript/src/tengi-core/com/noctarius/tengi/core/serialization/codec/Decoder.ts
module com.noctarius.core.serialization.codec {

    export interface Decoder {
        readBytes(bytes:Uint8Array): void;
        readBytes(bytes:Uint8Array, offset:number, length:number): void;
        readBoolean(): boolean;
        readBitSet(): Array<boolean>;
        readByte(): number;
        readUnsignedByte(): number;
        readShort(): number;
        readChar(): number;
        readInt32(): number;
        readCompressedInt32(): number;
        readInt64(): number;
        readCompressedInt64(): number;
        readFloat(): number;
        readDouble(): number;
        readString(): string;
        readObject(): any;
        readNullableObject(): any;
    }
}